### PR TITLE
add firefox to the list of secure browsers

### PR DESCRIPTION
**Pull request by cherti**
This PR adds Firefox to the list of secure browsers by only using internal configuration options.

### CODE REVIEW

**Comment by gautamkrishnar** [comment]
@cherti firefox was there on the list but was removed due to https://github.com/gautamkrishnar/nothing-private/issues/98 by @nunbit. 

@nunbit can you please verify this now.

**Comment by cherti** [comment]
According to the documentation this was added in 57 (as denoted in the file), however I did test this with 76.0.1, it would be interesting to get a firefox ESR-Test as well for older versions.

**Comment by cherti** [comment]
So, I have run some tests and the result seems to be: the Firefox-Fingerprint prevention works (even when simply reloading the same page), as long as the resolution of the browser window is different from where the form was originally filled out.
Only if it is *exactly* the same size the browser is recognized again.

This finding translates over different distributions, and, most importaintly, this also translates to the Tor Browser, which is currently listed as secure (and actually the behavior of reproducing windows sizes is one of its core features for the user's privacy).

**Comment by cherti** [comment]
Some further research has yielded this explanation for why Firefox adopted a different behavior for their fingerprinting: https://github.com/ghacksuserjs/ghacks-user.js/issues/767#issuecomment-515780193, they apparently decided to adapt the scheme of lowering entropy the Tor Browser already employs instead of raising it.

**Comment by kraliq233** [comment]
[]()

**Comment by gautamkrishnar** [comment]
I tried the latest version of Firefox, still, I can see some abnormalities. See the video below: 
https://streamable.com/q15mgj
I am using the latest version of firefox.

@cherti if you had gone through the implementation of firefox's fingerprint restriction, can you please tell me what all parameters are randomized or blocked to avoid fingerprinting?

**Comment by cherti** [comment]
Yes, and this is precisely *intended* behavior for the sake of tracking protection. I haven't gone throught the implementation itself, but the point here is that canvas randomization is easily detectable, you just run the canvas-fingerprint twice in your fingerprinting routine and the fingerprinter knows exactly that you are applying canvas randomization. That being said, I have tested that with the ungoogled chromium, the only randomized part there, as far as I tested, is the canvas randomization (and likely this is the same for palemoon, I'm extrapolating, but as enabling "canvas.poisondata" is required to make it fool this specific test this seems very likely even without testing).

Therefore, these fingerprinting protections are actually artifacts that can be used for fingerprinting by just slightly modifying the fingerprinting-Javascript (and potentially you can even do statistical attacks on the randomization itself, ungoogled chromium says "maximum 10 pixels", if palemoon does more they are, statistically, easily distinguishable here).

Therefore, the Firefox-approach, taken from the Tor Browser, is fundamentally different: the canvas-API will always return blank white, independent of where the Firefox is running. Instead of making every Firefox randomly unique, they decided to make every firefox look alike. Given that the Browser is extracted by the fingerprinting-JS as well, this is actually denying any information extracted from the canvas API, whereas, as described above, randomization gives away information to the tracking party.

The question now boils down to what you mean by "secure browser". If this this means "fails this specific test", then Firefox will only pass this test if a lot Firefox users use it at the same time (that being said, in this case I'd argue that "secure browsers" is a dangerously misleading claim). If it means "protects you from fingerprinting", then Firefox is actually better in this, because it deprives the tracking party of any additional tracking information by making all firefoxes look alike with regards to their canvases (and many other things that are tweaked by that option).

**Comment by nunbit** [comment]
> @cherti firefox was there on the list but was removed due to #98 by @nunbit.
> 
> @nunbit can you please verify this now.
> 
> We need more verifications from other firefox users.
NO. Nothing changed in 77. same failure, not even on Librewolf with hardened user.js.

**Comment by nunbit** [comment]
Also I am deleting my account due to Microsoft ownership. Consider moving to gitlab :). Bye bye.

**Comment by gautamkrishnar** [comment]
@cherti thank you for contribution, really sorry we cannot add firefox to the list since nothing private is working fine on default firefox, at least for me.

**Comment by Thorin-Oakenpants** [comment]
> canvas randomization is easily detectable, you just run the canvas-fingerprint twice in your fingerprinting routine

Not if the spoof is cached and re-used (for example see Canvas Blocker's default method). However, you can still detect if canvas is being altered in a separate test where the hash should be consistent over all browsers/devices

> Therefore, the Firefox-approach, taken from the Tor Browser, is fundamentally different: the canvas-API will always return blank white

In FF78+ the canvas is **randomized on each request** (not cached) and is pretty much unique (8 colors, specific pattern, dead give-away on size) - but RFP is not trying to hide itself, and all RFP users (such as Tor Browser users) would still all be the same (on randomizing metrics such as colors etc), as an enforced set.

---

I'm not sure what you guys are angling at with this list: Tor Browser/Firefox returning a white canvas or randomizing (78+) are both effective in rendering that FPing metric useless (even if scripts were to deep dive canvas randomizing characteristics), so I have no idea why you seem to think TB (or FF with RFP) aren't "browsers with countermeasures"
- If the script gets a white canvas, it only gets one value: "white canvas"
- If the script gets a random canvas and that's all they do, they get a poison pill
- If the script gets two different random canvasses on two passes, it still only gets one value: "random"
- If the script gets the same random canvasses on two passes (e.g. Canvas Blocker) and doesn't check further, they swallowed a poisoned pill
- If the script gets a random canvas but analyses it as being poisoned by e.g. RFP characteristics, they still only get one value: e.g. "random RFP"
- If the script does an extra canvas tampering detection test (and you allow for the RFP white canvas), they still only get one value: "random" (why else would canvas be tampered with)

In all those cases above, except where the script swallows the poison pill which screws up their FPing, all RFP/TB users are protected: they are still all the same.

You guys just need to up your game by detecting canvas mitigations: it's only when you swallow the poisoned data that you **think** it has no protection, but in reality it **is** protected. You can easily code to bypass being poisoned: which will leave you with a stable FP metric: and that stable FP metric will show that the protection is working (as all others will be the same: e.g. "random RFP" or "random extension" or "random palemoon")

tl;dr: stop taking the blue pill

**Comment by gautamkrishnar** [comment]
@Thorin-Oakenpants Thanks a lot for the explanation, @cherti had already opened #112 With firefox on the list.

**Comment by cherti** [comment]
> In FF78+ the canvas is randomized on each request (not cached) and is pretty much unique (8 colors, specific pattern, dead give-away on size) - but RFP is not trying to hide itself, and all RFP users (such as Tor Browser users) would still all be the same (on randomizing metrics such as colors etc), as an enforced set.

@Thorin-Oakenpants I've read about this, I wasn't aware that this would already land in FF78, thanks, that is indeed good to know. :)

**Comment by Thorin-Oakenpants** [comment]
just use https://ghacksuserjs.github.io/TorZillaPrint/TorZillaPrint.html#canvas to test stuff: it does two passes and can detect random on each request, it handles RFP as white canvas in FF77 or lower, it handles RFP random in FF78+, and it knows if RFP is on via another method, so can tell if random on each pass is by an extension. And I plan to add more information (such as colors used, file size, etc to make sure I get no false positives on RFP: not that I need it as RFP still doesn't randomize isPoint* methods, they're still static, but spoofed)

And was already digging into diving canvas randomizing: i.e detecting when it's spoofed, but stealth/cached - e.g. see https://thorin-oakenpants.github.io/testing/canvas.html

So anyway #112 will be out of date shortly: you'll want to add FF78+ and TB (based on ESR78: I can't quote TB version numbers because the change might be clean in stable, but not alpha, which is already on version 10) to the random list, and change them in the other list as well. BTW, it's not actually `Depriving websites of canvas data readout`: there is a difference in fingerprinting if something is blocked vs spoofing. It's just spoofing as white, so I would change that

**Comment by Thorin-Oakenpants** [comment]
> I can't quote TB version numbers because the change might be clean in stable, but not alpha, which is already on version 10

Just to be clear, TB stable/release major version numbers change when the ESR major number change
- first TB release based on ESR60 was version `8.0.0`
- first TB release based on ESR68 was version `9.0.0`
- first TB release based on ESR78 **will be** version `10.0.0`

but alpha changes mid-stream, and is ahead of stable
- first alpha version to be based on ESR68 was version `9.0a.6`
- current alpha `10.0a1` is still on ESR68

So if you want to use "TB10+ = random" and "TB9.5 or lower" = white canvas, just make sure to add that this numbering excludes alpha, and/or add in the ESR version it's based on

**Comment by cherti** [comment]
> BTW, it's not actually Depriving websites of canvas data readout: there is a difference in fingerprinting if something is blocked vs spoofing. It's just spoofing as white, so I would change that

Good point, done in https://github.com/gautamkrishnar/nothing-private/pull/113.

I have left Tor Browser as it is for now, once the new release is out it can be changed in the list accordingly and alpha and non-alpha do not have to be differentiated explicitly.
